security-proxy/src/main/java/org/georchestra/security/Proxy.java

```diff
@@ -369,49 +431,76 @@ private String buildForwardRequestURL(HttpServletRequest request) {
 
-            logger.debug("handlePathEncodedRequests: -- Handling Request: "+requestType+":"+forwardRequestURI+" from: "+request.getRemoteAddr());
+            logger.debug("handlePathEncodedRequests: -- Handling Request: " + requestType + ":" + forwardRequestURI + " from: " + request.getRemoteAddr());
 
             String sURL = findTarget(forwardRequestURI);
 
-            if(sURL == null){
+            if (sURL == null) {
                 response.sendError(404);
             }
 
             URL url;
             try {
                 url = new URL(sURL);
-            }catch (MalformedURLException e) {
-                throw new MalformedURLException(sURL +" is not a valid URL");
+            } catch (MalformedURLException e) {
+                throw new MalformedURLException(sURL + " is not a valid URL");
+            }
+
+            boolean sameHostAndPort = false;
+
+            try {
+                sameHostAndPort = isSameHostAndPort(request, url);
+            } catch (UnknownHostException e) {
+                logger.error("Unknown host in requested URL", e);
+                response.sendError(503);
+                return;
             }
 
-            if (isSameHostAndPort(request, url) && (isRecursiveCallToProxy(forwardRequestURI, contextPath)
-                    || isRecursiveCallToProxy(url.getPath(), contextPath))) {
+            if (sameHostAndPort && (isRecursiveCallToProxy(forwardRequestURI, contextPath) || isRecursiveCallToProxy(url.getPath(), contextPath))) {
                 response.sendError(403, forwardRequestURI + " is a recursive call to this service.  That is not a legal request");
             }
 
             if (request.getQueryString() != null && !isFormContentType(request)) {
                 StringBuilder query = new StringBuilder("?");
                 Enumeration paramNames = request.getParameterNames();
-                while(paramNames.hasMoreElements()) {
+                boolean needCasValidation = false;
+                while (paramNames.hasMoreElements()) {
                     String name = (String) paramNames.nextElement();
                     String[] values = request.getParameterValues(name);
                     for (String string : values) {
-                        if(query.length() > 1) {
+                        if (query.length() > 1) {
                             query.append('&');
                         }
-                        query.append(name);
-                        query.append('=');
-                        query.append(URLEncoder.encode(string, "UTF-8"));
+                        // special case: if we have a ticket parameter and no
+                        // authentication principal, we need to validate/open
+                        // the session against CAS server
+                        if ((request.getUserPrincipal() == null)
+                                && (name.equals(ServiceProperties.DEFAULT_CAS_ARTIFACT_PARAMETER))) {
```

Comment:
This adds a new feature here, namely we detect a CAS ticket here and try to open a session onto the SP.
Nice, but would have been better in a separated PR.